vault/resource_pki_backend_role.go

```diff
@@ -0,0 +1,380 @@
+				Optional:    true,
+				Computed:    true,
+				Description: "Specifies the O (Organization) values in the subject field of issued certificates.",
+			},
+			"generate_lease": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "Specifies if certificates issued/signed against this role will have Vault leases attached to them.",
+			},
+			"no_store": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "If set, certificates issued/signed against this role will not be stored in the storage backend.",
+			},
+			"backend": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Unique name of the PKI backend to configure.",
+				ForceNew:    true,
+				Default:     "pki",
+				// standardise on no beginning or trailing slashes
+				StateFunc: func(v interface{}) string {
+					return strings.Trim(v.(string), "/")
+				},
+			},
+		},
+	}
+}
+
+func pkiBackendRoleWrite(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	backend := d.Get("backend").(string)
+	role := d.Get("role").(string)
+
+	path := pkiBackendRolePath(backend, role)
+
+	log.Printf("[DEBUG] Writing PKI backend role %q", path)
+
+	data := map[string]interface{}{}
+
+	data["ttl"] = d.Get("ttl").(string)
+	data["max_ttl"] = d.Get("max_ttl").(string)
+	data["allow_localhost"] = d.Get("allow_localhost").(bool)
+
+	iDomains := d.Get("allowed_domains").([]interface{})
+	domains := make([]string, 0, len(iDomains))
+	for _, iDomain := range iDomains {
+		domains = append(domains, iDomain.(string))
+	}
+	if len(domains) > 0 {
+		data["allowed_domains"] = domains
+	}
+
+	data["allow_bare_domains"] = d.Get("allow_bare_domains").(bool)
+	data["allow_subdomains"] = d.Get("allow_subdomains").(bool)
+	data["allow_glob_domains"] = d.Get("allow_glob_domains").(bool)
+	data["allow_any_name"] = d.Get("allow_any_name").(bool)
+	data["enforce_hostnames"] = d.Get("enforce_hostnames").(bool)
+	data["allow_ip_sans"] = d.Get("allow_ip_sans").(bool)
+	data["client_flag"] = d.Get("client_flag").(bool)
+	data["code_signing_flag"] = d.Get("code_signing_flag").(bool)
+	data["email_protection_flag"] = d.Get("email_protection_flag").(bool)
+	data["key_type"] = d.Get("key_type").(string)
+	data["key_bits"] = d.Get("key_bits").(int)
+
+	iUsages := d.Get("key_usage").([]interface{})
+	usages := make([]string, 0, len(iUsages))
+	for _, iUsage := range iUsages {
+		usages = append(usages, iUsage.(string))
+	}
+	if len(usages) > 0 {
+		data["key_usage"] = usages
+	}
+
+	data["use_csr_common_name"] = d.Get("use_csr_common_name").(bool)
+	data["use_csr_sans"] = d.Get("use_csr_sans").(bool)
+	data["ou"] = d.Get("ou").(string)
+	data["organization"] = d.Get("organization").(string)
+	data["generate_lease"] = d.Get("generate_lease").(bool)
+	data["no_store"] = d.Get("no_store").(bool)
+
+	_, err := client.Logical().Write(path, data)
+
+	d.SetId(path)
```

Comment:
You probably picked this up looking at the pre-existing code, but I think we're trying to move to using `d.SetId` after checking the error, which means we don't need to call `SetId` in the error handling code.

Reply:
Makes sense. Done.